File: ruby/shegerpay.rb
# frozen_string_literal: true

# ShegerPay Ruby SDK
# Official Ruby SDK for ShegerPay Payment Verification Gateway
#
# @author ShegerPay <support@shegerpay.com>
# @version 1.0.0

require 'net/http'
require 'uri'
require 'json'
require 'openssl'

module ShegerPay
  VERSION = '1.0.0'
  
  class Error < StandardError; end
  class AuthenticationError < Error; end
  class ValidationError < Error; end
  
  # Verification result
  class VerificationResult
    attr_reader :verified, :valid, :status, :provider, :transaction_id, :amount, :reason, :mode
    
    def initialize(data)
      @verified = data['verified'] || data['valid'] || false
      @valid = data['valid'] || false
      @status = data['status'] || 'unknown'
      @provider = data['provider']
      @transaction_id = data['transaction_id']
      @amount = data['amount']
      @reason = data['reason']
      @mode = data['mode']
    end
    
    def valid?
      @valid
    end
  end
  
  # ShegerPay Payment Verification Client
  class Client
    DEFAULT_BASE_URL = 'https://api.shegerpay.com'
    
    # Create a new ShegerPay client
    #
    # @param api_key [String] Your secret API key (sk_test_xxx or sk_live_xxx)
    # @param options [Hash] Optional configuration
    def initialize(api_key, options = {})
      raise AuthenticationError, 'API key is required' if api_key.nil? || api_key.empty?
      
      unless api_key.start_with?('sk_test_', 'sk_live_')
        raise AuthenticationError, 'Invalid API key format'
      end
      
      @api_key = api_key
      @base_url = (options[:base_url] || DEFAULT_BASE_URL).chomp('/')
      @timeout = options[:timeout] || 30
      @mode = api_key.start_with?('sk_test_') ? 'test' : 'live'
    end
    
    # Verify a payment transaction
    #
    # @param params [Hash] Verification parameters
    # @return [VerificationResult]
    def verify(params)
      transaction_id = params[:transaction_id]
      amount = params[:amount]
      provider = params[:provider]
      merchant_name = params[:merchant_name] || 'ShegerPay Verification'
      sender_account = params[:sender_account]
      
      raise ValidationError, 'transaction_id is required' unless transaction_id
      raise ValidationError, 'amount is required' unless amount
      
      provider ||= transaction_id.downcase.include?('cs.bankofabyssinia.com/slip/?trx=') ? 'boa' : nil
      raise ValidationError, 'provider is required for ambiguous transaction references. Pass provider explicitly or use quick_verify.' unless provider
      
      data = {
        provider: provider,
        transaction_id: transaction_id,
        amount: amount,
        merchant_name: merchant_name
      }
      data[:sub_provider] = params[:sub_provider] if params[:sub_provider]
      data[:sender_account] = sender_account if sender_account
      
      response = request(:post, '/api/v1/verify', data)
      VerificationResult.new(response)
    end
    
    # Quick verification with auto-detected provider
    #
    # @param transaction_id [String] Bank transaction reference
    # @param amount [Float] Expected amount
    # @return [VerificationResult]
    def quick_verify(transaction_id, amount, expected_provider = nil, sender_account = nil)
      payload = {
        transaction_id: transaction_id,
        amount: amount
      }
      payload[:expected_provider] = expected_provider if expected_provider
      payload[:sender_account] = sender_account if sender_account
      response = request(:post, '/api/v1/quick-verify', payload)
      VerificationResult.new(response)
    end
    
    # Get transaction history
    #
    # @param limit [Integer] Maximum number of transactions
    # @return [Array]
    def history(limit = 50)
      request(:get, '/api/v1/history')
    end
    
    # Verify webhook signature
    #
    # @param payload [String] Raw request body
    # @param signature [String] X-ShegerPay-Signature header
    # @param secret [String] Your webhook secret
    # @return [Boolean]
    def self.verify_webhook_signature(payload, signature, secret)
      expected = 'sha256=' + OpenSSL::HMAC.hexdigest('SHA256', secret, payload)
      Rack::Utils.secure_compare(expected, signature)
    rescue
      expected == signature
    end
    
    private
    
    def request(method, path, data = nil)
      uri = URI.parse("#{@base_url}#{path}")
      
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == 'https'
      http.open_timeout = @timeout
      http.read_timeout = @timeout
      
      case method
      when :get
        request = Net::HTTP::Get.new(uri)
      when :post
        request = Net::HTTP::Post.new(uri)
        request.body = URI.encode_www_form(data) if data
        request['Content-Type'] = 'application/x-www-form-urlencoded'
      end
      
      request['X-API-Key'] = @api_key
      request['User-Agent'] = 'ShegerPay-Ruby-SDK/1.0'
      
      response = http.request(request)
      
      case response.code.to_i
      when 401
        raise AuthenticationError, 'Invalid API key'
      when 400
        error = JSON.parse(response.body) rescue {}
        raise ValidationError, error['detail'] || 'Validation error'
      end
      
      JSON.parse(response.body) rescue {}
    end
  end
end

File: swift/ShegerPay.swift
/**
 * ShegerPay Swift SDK
 * Official Swift SDK for ShegerPay Payment Verification Gateway
 * 
 * Usage:
 *   let client = ShegerPay(apiKey: "sk_test_xxx")
 *   let result = try await client.verify(transactionId: "FT123456", amount: 100, provider: "cbe")
 */

import Foundation
import CommonCrypto

// MARK: - Errors

public enum ShegerPayError: Error, LocalizedError {
    case invalidApiKey
    case missingApiKey
    case authenticationFailed
    case validationError(String)
    case networkError(Error)
    case invalidResponse
    
    public var errorDescription: String? {
        switch self {
        case .invalidApiKey: return "Invalid API key format"
        case .missingApiKey: return "API key is required"
        case .authenticationFailed: return "Authentication failed"
        case .validationError(let msg): return msg
        case .networkError(let error): return error.localizedDescription
        case .invalidResponse: return "Invalid response from server"
        }
    }
}

// MARK: - Models

public struct VerificationResult: Codable {
    public let verified: Bool?
    public let valid: Bool
    public let status: String
    public let provider: String?
    public let transactionId: String?
    public let amount: Double?
    public let reason: String?
    public let mode: String?
    public let payer: String?
    
    enum CodingKeys: String, CodingKey {
        case verified, valid, status, provider, amount, reason, mode, payer
        case transactionId = "transaction_id"
    }
}

public struct PaymentLink: Codable {
    public let id: String
    public let shortCode: String
    public let paymentUrl: String
    public let qrCodeBase64: String
    public let status: String
    
    enum CodingKeys: String, CodingKey {
        case id, status
        case shortCode = "short_code"
        case paymentUrl = "payment_url"
        case qrCodeBase64 = "qr_code_base64"
    }
}

// MARK: - ShegerPay Client

public class ShegerPay {
    private let apiKey: String
    private let baseURL: String
    private let mode: String
    private let session: URLSession
    
    public init(apiKey: String, baseURL: String = "https://api.shegerpay.com") throws {
        guard !apiKey.isEmpty else {
            throw ShegerPayError.missingApiKey
        }
        
        guard apiKey.hasPrefix("sk_test_") || apiKey.hasPrefix("sk_live_") else {
            throw ShegerPayError.invalidApiKey
        }
        
        self.apiKey = apiKey
        self.baseURL = baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        self.mode = apiKey.hasPrefix("sk_test_") ? "test" : "live"
        
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config)
    }
    
    // MARK: - Verification
    
    /// Verify a payment transaction
    public func verify(
        transactionId: String,
        amount: Double,
        provider: String? = nil,
        merchantName: String? = nil,
        senderAccount: String? = nil
    ) async throws -> VerificationResult {
        let detectedProvider = provider ?? (transactionId.lowercased().contains("cs.bankofabyssinia.com/slip/?trx=") ? "boa" : nil)
        guard let detectedProvider else {
            throw ShegerPayError.validationError("provider is required for ambiguous transaction references. Pass provider explicitly or use quickVerify().")
        }
        
        var params: [String: String] = [
            "provider": detectedProvider,
            "transaction_id": transactionId,
            "amount": String(amount),
            "merchant_name": merchantName ?? "ShegerPay Verification"
        ]
        if let senderAccount, !senderAccount.isEmpty {
            params["sender_account"] = senderAccount
        }
        
        return try await request(method: "POST", path: "/api/v1/verify", params: params)
    }
    
    /// Quick verification with auto-detected provider
    public func quickVerify(
        transactionId: String,
        amount: Double,
        expectedProvider: String? = nil,
        senderAccount: String? = nil
    ) async throws -> VerificationResult {
        var params: [String: String] = [
            "transaction_id": transactionId,
            "amount": String(amount)
        ]
        if let expectedProvider, !expectedProvider.isEmpty {
            params["expected_provider"] = expectedProvider
        }
        if let senderAccount, !senderAccount.isEmpty {
            params["sender_account"] = senderAccount
        }
        return try await request(method: "POST", path: "/api/v1/quick-verify", params: params)
    }
    
    // MARK: - Payment Links
    
    /// Create a payment link
    public func createPaymentLink(
        title: String,
        amount: Double,
        currency: String = "ETB",
        description: String? = nil
    ) async throws -> PaymentLink {
        var params: [String: Any] = [
            "title": title,
            "amount": amount,
            "currency": currency,
            "enable_cbe": true,
            "enable_telebirr": true
        ]
        
        if let desc = description {
            params["description"] = desc
        }
        
        return try await requestJSON(method: "POST", path: "/api/v1/payment-links/", json: params)
    }
    
    // MARK: - Private Methods
    
    private func request<T: Decodable>(
        method: String,
        path: String,
        params: [String: String]
    ) async throws -> T {
        var urlComponents = URLComponents(string: baseURL + path)!
        
        var request = URLRequest(url: urlComponents.url!)
        request.httpMethod = method
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.setValue("ShegerPay-Swift-SDK/1.0", forHTTPHeaderField: "User-Agent")
        
        if method == "POST" {
            let body = params.map { "\($0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? $0.value)" }
                .joined(separator: "&")
            request.httpBody = body.data(using: .utf8)
        }
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ShegerPayError.invalidResponse
        }
        
        if httpResponse.statusCode == 401 {
            throw ShegerPayError.authenticationFailed
        }
        
        return try JSONDecoder().decode(T.self, from: data)
    }
    
    private func requestJSON<T: Decodable>(
        method: String,
        path: String,
        json: [String: Any]
    ) async throws -> T {
        var request = URLRequest(url: URL(string: baseURL + path)!)
        request.httpMethod = method
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: json)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ShegerPayError.invalidResponse
        }
        
        if httpResponse.statusCode == 401 {
            throw ShegerPayError.authenticationFailed
        }
        
        return try JSONDecoder().decode(T.self, from: data)
    }
    
    // MARK: - Webhook Verification
    
    /// Verify webhook signature
    public static func verifyWebhookSignature(payload: String, signature: String, secret: String) -> Bool {
        guard let keyData = secret.data(using: .utf8),
              let payloadData = payload.data(using: .utf8) else {
            return false
        }
        
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        
        keyData.withUnsafeBytes { keyBytes in
            payloadData.withUnsafeBytes { dataBytes in
                CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA256),
                       keyBytes.baseAddress, keyData.count,
                       dataBytes.baseAddress, payloadData.count,
                       &digest)
            }
        }
        
        let expected = "sha256=" + digest.map { String(format: "%02x", $0) }.joined()
        return expected == signature
    }
}
